RFC-0022-Model-Context-Protocol.md

```diff
@@ -0,0 +1,127 @@
+# **RFC-0022 for Presto**
+
+## Presto Model Context Protocol Support
+
+Proposers
+
+* Reetika Agrawal
+
+## Summary
+
+This proposal introduces support for integrating Model Context Protocol (MCP) with Presto through a separate lightweight process, referred to as the MCP Server.
+The MCP Server acts as a protocol translation layer between AI agents using JSON-RPC and Presto’s existing HTTP-based query protocol.
+
+## Background
+
+Presto exposes a RESTful HTTP API for query submission and result retrieval. Clients submit a query via a POST request and receive a tokenized nextUri chain for incremental result fetching.
+However, AI frameworks like OpenAI MCP operate on a request–response model over JSON-RPC, where clients expect complete results in a single response, not streamed batches.
+Because MCP and Presto differ fundamentally in communication models, a direct integration is impractical. Additionally, direct embedding of MCP into Presto coordinators would disrupt routing and proxy layers that rely on Presto’s native HTTP semantics.
+
+### Goals
+
+- Enable MCP-compatible AI agents (e.g., OpenAI ChatGPT tools) to query Presto seamlessly.
+- Preserve all existing Presto router, proxy, and load-balancing infrastructure without modification.
+- Simplify authentication by reusing existing OAuth/JWT mechanisms.
+- Prevent unbounded queries by applying automatic limits when appropriate.
+
+### Proposed Plan
+
+Introduce a new lightweight service: presto-mcp-server, deployed alongside Presto coordinators and Presto Router.
+
+The MCP server will:
+
+- Expose a JSON-RPC 2.0 HTTP endpoint (`/mcp`, `/v1/mcp`) interface to AI clients.
+- Implement the core MCP primitives:
+  - `tools/list` for tool discovery 
+  - `tools/call` for executing tools
+  - Initially provide a single tool: `query.run`, which executes SQL queries against Presto.
```

Comment:
Any restriction on the type of queries supported?  If it's a DML, might we need a new tool to support it, since they don't return data but return metadata about what was modified?